### PR TITLE
Feat: load address from peer

### PR DESCRIPTION
**Pull request by sharafian**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -88,25 +91,63 @@ class PluginXrpPaychan extends PluginBtp {
...
+
+    if (!this._incomingChannel) {
+      this._reloadingChannel = false
+      this._log.debug(`setting incoming channel to`, newId)
+      this._log.trace(`channel details are`, details)
+      this._incomingChannel = newId
+      this._incomingChannelDetails = details
+      this._lastClaimedAmount = new BigNumber(this.xrpToBase(this._incomingChannelDetails.balance))
+      this._incomingClaim = { amount: this._lastClaimedAmount.toString() }
+      this._lastClaimedAmount = new BigNumber(this.xrpToBase(this._incomingChannelDetails.balance))
```
Comment (sentientwaffle) [inline index.js]: Duplicate line (see 144)

File: index.js [hunk 2]
```diff
@@ -126,17 +167,34 @@ class PluginXrpPaychan extends PluginBtp {
...
     }
 
+    if (protocolMap.xrp_address) {
+      this._log.debug('got xrp_address request from peer')
+      this._setPeerAddress(protocolMap.xrp_address)
+
+      return [{
+        protocolName: 'xrp_address',
+        contentType: BtpPacket.MIME_TEXT_PLAIN_UTF8,
+        data: Buffer.from(this._address || '')
```
Comment (sentientwaffle) [inline index.js]: Why `|| ''`? Shouldn't `this._address` always be defined?

File: index.js [hunk 3]
```diff
@@ -88,25 +91,63 @@ class PluginXrpPaychan extends PluginBtp {
...
+    this._reloadingChannel = true
+
+    this._log.trace('validating new paychan. id=' + newId)
+    const details = await this._api.getPaymentChannel(newId)
+    this._validateChannelDetails(details)
+
+    this._log.trace('checking that peer\'s scale matches')
+    await this._getPeerInfo()
+
+    if (this._incomingChannel && newId !== this._incomingChannel) {
```
Comment (sentientwaffle) [inline index.js]: `newId !== this._incomingChannel` was checked at the beginning of this method.

File: index.js [hunk 4]
```diff
@@ -88,25 +86,63 @@ class PluginXrpPaychan extends PluginBtp {
...
+
+    this._log.trace('validating new paychan. id=' + newId)
+    const details = await this._api.getPaymentChannel(newId)
+    this._validateChannelDetails(details)
+
+    this._log.trace('checking that peer\'s scale matches')
+    await this._getPeerInfo()
+
+    // re-check in case of race conditions
+    if (this._incomingChannel && newId !== this._incomingChannel) {
```
Comment (d1no007) [inline index.js]: is the idea here behind these race conditions that the _setIncomingChannel function changes the _incomingChannel instance variable every time it is run, so we need to continue to check it's state after any async call?

Comment (sharafian) [reply index.js]: exactly, yeah

File: index.js [hunk 5]
```diff
@@ -312,6 +379,24 @@ class PluginXrpPaychan extends PluginBtp {
       await this._reloadIncomingChannelDetails()
     }
 
+    if (!this._peerAddress) {
```
Comment (d1no007) [inline index.js]: this _connect function overrides the custom _connect in BTP to send the xrp_address subprotocol after all normal BTP listeners are registered, right?

Comment (sharafian) [reply index.js]: yeah

**Pull request merged**